### DISCUSSION TITLE
issue: Watch field is not updated after using remove() from useFieldArray

### DISCUSSION THREAD

**Discussion by andre-ph** [react-hook-form/react-hook-form#10309] [Q&A] [answered]
### Version Number

7.43.9

### Codesandbox/Expo snack

https://codesandbox.io/s/agitated-darwin-6j8fje?file=/src/App.js

### Steps to reproduce

1. Go to page
2. Click on "+" button to add an item
3. An item appears
4. Click on "x" button on item row to remove it
5. The item is still here, check console to see that fields is update but not watchFieldArray

### Expected behaviour

The item should not be appearing anymore.

I suspected this because I filter the fields in a state and create an **originalIndex.** field on each item to keep track of the item original index to remove it.
This seems unnecessary in this sandbox but I need this filtering in my project.
Maybe this is something not possible with the RHF ?

I also put all the other things like I use in my project: yup schema, material ui, use of Controller but I don't think this is an issue with these, as they seem fine before I started to implement the filtering.

### What browsers are you seeing the problem on?

Firefox

### Relevant log output

```shell
Before clicking on remove button :
console.log("fields watch", fields, watchFieldArray); => fields watch (1) [Object] (1) [Object]
After clicking on remove button:
console.log("fields watch", fields, watchFieldArray); => fields watch [] (1) [Object]
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
Remove all side effects/dependencies and provide an isolated codesandbox, try to reproduce in this codesandbox: https://codesandbox.io/s/react-hook-form-usefieldarray-ssugn

**Comment by andre-ph** [comment]
Ok sorry for that I didn't know about the sandbox, here you are : https://codesandbox.io/s/react-hook-form-usefieldarray-forked-m4rxsf?file=/src/index.tsx

~~So I guess the problem doesn't come from the filtering as I expected but from somewhere else in my setup of RHF.~~
_edit:_ WatchFieldArray is not updated while fields is, when deleting a row after filtering

**Comment by leapful** [comment]
@andre-ph 

Could you please explain why you need to use `useWatch` hook instead of using just `fields` from `useFieldArray` hook and `getValues` method of `useForm`.

For example:

```
const { fields } = useFieldArray({ name: 'test' })

useEffect(() => {
   setFiltered(getValues('test').filter())
}, [fields)
```

**Reply by andre-ph** [reply]
I wanted to mimic the **Controlled Field Array** in the documentation.
Plus I did break my form into many components, I found it more convenient to use UseWatch instead of passing getValues in the props of my children components,  since getValues has to be taken from useForm in the parent component.

But you are right, it seems using getValues will most likely solve my issue, as it is indeed updated unlike useWatch.
I will try to implement it in my project, thanks for your help

**Reply by andre-ph** [reply]
Ok I have tested and getValues doesn't seem to be updated as well.

I have logged it in the Sandbox if you want to try.

You can observe this easily as:
- delete an item
- append an item
- a new item is appended & the previous item reappears

**Reply by leapful** [reply]
@andre-ph 

There are several issues with your implementation:

1. `fields` object from `useFieldArray` only contains default value and unique key to help simplify field array , it should not be used for mapping data.
2. `getValues` method already return latest data of the field array, you just need to filter on that instead of having complex redundant merging and filtering with `fields` object.

Just use `getValues` to filter is enough to solve the issue.

Working demo: https://codesandbox.io/s/react-hook-form-usefieldarray-forked-j04hd0?file=/src/index.tsx

**Reply by andre-ph** [reply]
@leapful 

Your demo is not working, 
If you delete an item, it doesn't get removed at all.
**Plus if you print getValues, it doesn't get updated.**

It's surely because the ids that come from fields are not passed to the filteredFields state, hence I mapped on the fields value in the first place.

Maybe it's wrong to do this but I got this part from the doc : https://react-hook-form.com/api/usefieldarray/ (**ControlledFieldArray** part), maybe it's not up to date ?

**Reply by leapful** [reply, marked as answer]
@andre-ph 

Sorry, my example codesandbox has been overridden by auto save on old tab. Here is the correct version with some optimizations.

https://codesandbox.io/s/react-hook-form-usefieldarray-forked-cksl5y?file=/src/index.tsx

The main idea is to get rid of altering `fields` object when using with `useFieldArray` hook, just does not render array item if it is not expected following filter criteria.

**Reply by andre-ph** [reply]
@leapful 
Oh no worries, I was wondering what changed lol

Now it works perfectly !
I tried to implement it in my project (and modifying a bit to adjust to my components) and it looks to work as well.
Now I will remember to avoid altering fields object for my future uses.

Thank you for your help and your time !
